agent/src/main/java/com/appland/appmap/record/RecordingSession.java

```diff
@@ -31,6 +31,22 @@ public Recorder.Metadata getMetadata() {
   }
 
   public synchronized void add(Event event) {
+    // Don't write pending events.
+    if (event.pending()) {
+      return;
+    }
+
+    // If there's a pending event associated with this event, though, it's time to write it out. 
+    Event pendingEvent = event.pendingEvent();
+    // Neither should still be marked pending, though:
+    assert (pendingEvent ==  null || !pendingEvent.pending()); assert !event.pending();
+    if (pendingEvent != null) {
+      writeEvent(pendingEvent);
```

Comment:
So if I get this right, events may be written out of sequence to the stream, but the ids will be correct. So, we should be sure and verify downstream that event sequence is not trusted in some way - that appmap-js always sequences events by id and not by the order of their appearance in the AppMap file.

Reply:
> So if I get this right, events may be written out of sequence to the stream, but the ids will be correct.
Yes, that's right.

> So, we should be sure and verify downstream that event sequence is not trusted in some way
Yeah, I wondered about that. It would be good to add a test to both appmap-js and appland-server to ensure that events that are out of order in the `events` array get interpreted correctly.

Reply:
Nowadays we don't use the old Golang appland uploader any more - everything goes through the JS uploader. If we ever bring back Mapsets upload, we should be sure and role that into `@appland/appmap`. So I think all we need to do is make sure that the JS AppMap uploader does the right thing and normalizes the events properly before uploading them. The server can/should/will assume that uploaded events have been normalized. This is a bit risky, but server side normalization was very slow, and duplicative with the JS code, and we've taken it out.

If we re-implement the server-side of upload in JS/TS, then we can add it back in as a safety measure.

Reply:
As far as I can tell, the uploader only normalizes/transforms a recording as part of pruning it -- if the AppMap is small enough, if gets uploaded unmodified. (Prior to my changes to add pruning before uploading, it was always uploaded verbatim.)

My reading of https://github.com/applandinc/appmap-js/blob/main/packages/models/src/appMapBuilder/eventSorter.js#L21 and https://github.com/applandinc/appmap-js/blob/main/packages/models/src/appMapBuilder/index.js#L56 is that the model doesn't depend on the order of elements in the `events` array. I believe that this means that any diagrams that show events will render correctly.

WDYT needs to be done?

Reply:
So, it looks to me like as the events are read, they are [pushed onto the stack for their thread](https://github.com/applandinc/appmap-js/blob/main/packages/models/src/appMapBuilder/eventStack.js#L17), and [added to a list in order of apperance](https://github.com/applandinc/appmap-js/blob/main/packages/models/src/appMapBuilder/eventStack.js#L17). So, the behavior of the Java agent is that if event `1` is an http_server_request, events `2..N-1` are function calls, and event `N` is the http_server_response, the events list will contain `[ 2..N-1, 1, N]`. Is that correct? It seems like the stack and the event list will have events out of order. I definitely know that the scanner assumes that the events in the event list are sorted in the order in which they occur. 

In any case, I think it would be a good idea to add a test case for appmap-js that to load one of these AppMaps from the Spring recorder and make sure that everything is kosher. Running a scan on such an AppMap to ensure that it's well-behaved also seems like a good idea. Unless I'm mis-understanding what the written event stream will look like.

Reply:
The simplest thing for the appmap-js AppMap loader to do would probably be just to sort all the events by id before doing anything else.

Reply:
Ok, you've convinced me that writing the events out of order isn't a good idea. Requiring the the uploader sort all the events in every AppMap seems very expensive.

I'll look for a way to have the Java agent write the events in the proper order.